db/dcrpg/upgrades.go

```diff
@@ -242,10 +266,132 @@ func compatVersion1Upgrades(db *sql.DB, current, target DatabaseVersion, _ Block
+		}
+		_, err = u.db.Exec(`ALTER TABLE blocks ADD COLUMN IF NOT EXISTS winners TEXT[];`)
+		if err != nil {
+			return fmt.Errorf("Add winners column error: %v", err)
+		}
+	}
+	// Do everything else under a transaction.
+	dbTx, err := u.db.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to create db transaction: %v", err)
+	}
+	defer func() {
+		if err == nil {
+			dbTx.Commit()
+		} else {
+			dbTx.Rollback()
+		}
+	}()
+	makeErr := func(s string, args ...interface{}) error {
+		err = fmt.Errorf(s, args...)
+		return err
+	}
+	// Start with a height-ordered list of block data.
+	blockRows, err := u.db.Query(`
+		SELECT id, hash, height
+		FROM blocks
+		WHERE is_mainchain
+		ORDER BY height
+	;`)
+	if err != nil {
+		return makeErr("block hash query error: %v", err)
+	}
+	defer blockRows.Close()
+	// Set the stake database to the genesis block.
+	dir, err := ioutil.TempDir("", "tempstake")
+	if err != nil {
+		return makeErr("unable to create temp directory")
+	}
+	defer os.RemoveAll(dir)
+	sDB, _, err := u.stakeDB.EmptyCopy(dir)
+	if err != nil {
+		return makeErr("stake db init error: %v", err)
+	}
+	// Two prepared statements.
+	statsStmt, err := dbTx.Prepare(internal.UpsertStats)
+	if err != nil {
+		return makeErr("failed to prepare stats insert statement: %v", err)
+	}
+	// sql does not deal with PostgreSQL array syntax, it must be Sprintf'd.
+	winnersStmt, err := dbTx.Prepare("UPDATE blocks SET winners = $1 where hash = $2;")
+	if err != nil {
+		return makeErr("failed to prepare winners insert statement: %v", err)
+	}
+
+	checkHeight := 0
+	var hashStr string
+	var id, height int
+	for blockRows.Next() {
+		if u.ctx.Err() != nil {
+			return makeErr("context cancelled. rolling back update")
+		}
+		blockRows.Scan(&id, &hashStr, &height)
+		hash, err := chainhash.NewHashFromStr(hashStr)
+		if err != nil {
+			return makeErr("NewHashFromStr: %v", err)
+		}
+		// If the height is not the expected height, the database must be corrupted.
+		if height != checkHeight {
+			return makeErr("height mismatch %d != %d. database corrupted!", height, checkHeight)
+		}
+		checkHeight += 1
+		// A periodic update messaage.
+		if height%10000 == 0 {
+			log.Infof("Processing blocks %d - %d", height, height+9999)
+		}
+		// Connecting the block updates the live ticket cache and ticket info cache.
+		// The StakeDatabase is pre-populated with the genesis block, so skip it.
+		if height > 0 {
+			_, err = sDB.ConnectBlockHash(hash)
+			if err != nil {
+				return makeErr("ConnectBlockHash: %v", err)
+			}
+		}
+
+		// The "best" pool info is for the chain at the tip just added.
+		poolInfo := sDB.PoolInfoBest()
+		if poolInfo == nil {
+			return makeErr("PoolInfoBest error encountered")
+		}
+		// Insert rows.
+		_, err = statsStmt.Exec(id, height, poolInfo.Size, int64(poolInfo.Value*dcrToAtoms))
+		if err != nil {
+			return makeErr("insert Exec: %v", err)
+		}
+		_, err = winnersStmt.Exec(pq.Array(poolInfo.Winners), hashStr)
```

Comment:
Looks like you got the pq.Array solution here, but the comment above on line 345.

BTW, I believe you can be more explicit with `pq.StringArray`, but [pq sorts it out with a type switch](https://github.com/lib/pq/blob/78223426e7c66d631117c0a9da1b7f3fde4d23a5/array.go#L41).

Reply:
`InsertBlock` was doing the sprintf with some other text arrays (and uint arrays), so perhaps I'm mistaken about something.